client/extensions/woocommerce/components/product-search/row.js

```diff
@@ -0,0 +1,295 @@
+			variations: [],
+		};
+	}
+
+	componentDidMount() {
+		const { siteId, product, productId } = this.props;
+
+		if ( siteId && productId ) {
+			if ( ! product || 'variable' !== product.type ) {
+				return;
+			}
+			this.props.fetchProductVariations( siteId, productId );
+		}
+	}
+
+	componentWillReceiveProps( newProps ) {
+		// Short out if we know the new product doesn't need variations
+		if ( ! newProps.product || 'variable' !== newProps.product.type ) {
+			return;
+		}
+
+		const { oldSiteId, oldProductId } = this.props;
+		const { newSiteId, newProductId } = newProps;
+
+		if ( oldSiteId !== newSiteId || oldProductId !== newProductId ) {
+			this.props.fetchProducts( newSiteId, newProductId );
+		}
+
+		const selectedIds = intersection( newProps.value, newProps.product.variations );
+		const selectedVariations = selectedIds.map( id => find( newProps.variations, { id } ) );
+		this.state = {
+			showForm: Boolean( selectedIds.length ),
+			variations: filter( selectedVariations ) || [],
+		};
+	}
+
+	isSelected = id => isProductSelected( this.props.value, id );
+
+	onChange = event => {
+		const productId = Number( event.target.value );
+		this.setState(
+			prevState => ( {
+				variations: filter( prevState.variations, item => item.id !== productId ),
+			} ),
+			() => {
+				this.props.onChange( productId );
+			}
+		);
+	};
+
+	toggleCustomizeForm = event => {
+		// This handler can be on the label, or button with the label, so we
+		// stop propagation to avoid immediate open-then-close behavior.
+		event.stopPropagation();
+		this.setState( prevState => {
+			// Open form
+			if ( ! prevState.showForm ) {
+				return { showForm: true };
+			}
+			// Only close form if nothing's selected
+			if ( prevState.showForm && ! this.areAnySelected() ) {
+				return { showForm: false };
+			}
+		} );
+	};
+
+	updateItem = attributes => {
+		const { variations } = this.props;
+		// Don't swap the product if we have an "any" selected
+		if ( -1 !== values( attributes ).indexOf( 'any' ) ) {
+			return;
+		}
+		// Using filter instead of find to make sure we find exactly one match.
+		const matchingVariations = filter( variations, v => {
+			return reduce(
+				v.attributes,
+				( result, a ) => {
+					return result && attributes[ a.name ] === a.option;
+				},
+				true
+			);
+		} );
+		if ( matchingVariations.length === 1 ) {
+			// We found a match.
+			const variation = head( matchingVariations );
+			if ( this.props.singular ) {
+				this.setState( { variations: [ variation ] } );
+				this.props.onChange( variation.id );
+				return;
+			}
+			this.setState(
+				prevState => ( {
+					variations: uniqBy( [ ...prevState.variations, variation ], 'id' ),
+				} ),
+				() => {
+					forEach( this.state.variations, v => this.props.onChange( v.id ) );
+				}
+			);
+		}
+	};
+
+	areAnySelected = () => {
+		const { product, singular } = this.props;
+		const { variations } = this.state;
+		if ( singular && variations[ 0 ] ) {
```

Comment:
No to both, really, because of the line after both of these; behavior following is different for singular vs multiple selections. If you want to take a stab at updating this so that it's not the case once it's merged, feel free :)